draft-normington-jsonpath-latest.txt

```diff
@@ -524,73 +524,100 @@ Internet-Draft   JavaScript Object Notation (JSON) Path     October 2020
+   If "step > 0", then "start" defaults to "0" while "end" defaults to
+   "len".
 
-   For positive integer "step" and integers "start" and "end", the array
-   slice "start:end:step" consists of the indices starting at "start"
-   and increasing repeatedly by "step" while the resultant index is less
-   than "end".  The indices of the slice are the same as the values of i
-   iterated over by this "for" loop in C:
+   If "step < 0", then "start" defaults to "len-1", while an omitted
+   "end" causes the iteration to terminate at 0 (inclusive).
 
+   If "step" is "0", the resulting array is empty.
 
+   If "step" is omitted, it defaults to "1".
 
+   The resulting array is then constructed by iterating from "start"
+   (inclusive) to "end" (exclusive), collecting array elements at index
+   "i", and incrementing the iterator "i" by "step" after each
+   iteration.
 
 
 
-Normington & Surov        Expires 10 April 2021                [Page 10]
+
+Normington & Surov        Expires 19 April 2021                [Page 10]
 
 Internet-Draft   JavaScript Object Notation (JSON) Path     October 2020
 
 
+   If "len" is "0" the resulting array is empty.
+
+   If "step > 0" and "start > end" the resulting array is empty.
+
+   If "step < 0" and "start < end" the resulting array is empty.
+
+Pseudocode
+
+   If "step > 0", the indices of the slice are the same as the values of
+   "i" iterated over by this for loop in C:
+
    for (i = start; i < end; i = i + step) {
      ...
    }
 
-   When "step" is positive, the default value of "start" is "0" and the
-   default value of "end" is the length of the array.
-
-   For negative integer "step" and integers "start" and "end", the array
-   slice "start:end:step" consists of the indices starting at "start"
-   and decreasing repeatedly by "step" while the resultant index is
-   greater than "end".  The indices of the slice are the same as the
-   values of i iterated over by this "for" loop in C:
+   If "step < 0", the iteration will proceed backwards and the iteration
+   test must be reversed:
 
    for (i = start; i > end; i = i + step) {
      ...
    }
 
-   When "step" is negative, the default value of "start" is "len - 1"
-   (where "len" is the length of the array) and the default value of
-   "end" is "-1" (literally, not in the earlier sense of a shorthand for
-   "len - 1").
+   If the "end" bound is omitted, the iteration must terminate at 0
+   inclusive.  This can be achieved by setting "end = -1", if the
+   underyling numeric type allows it.  Alternatively the termination
+   condition can be changed to "i >= 0".
+
+Limits
+
+   The RECOMMENDED maximum length of the array is 2^64.  The numerical
+   domain of array indices is thus "0 <= b < 2^64".
 
-   For zero "step", the array slice is empty and no elements are
-   selected.
+   The RECOMMENDED domain of input slice bounds is "-2^64 <= b < 2^64".
+   If not possible, implementations SHOULD at least aim for "-2^63 <= b
+   < 2^63".
```

Comment:
I'm not a fan of specifying a limit, though I do agree with the usage of RECOMMENDED to allow leeway for implementations.

Instead of specifying a hard limit, I think we should specify a behavior when an implementation encounters values outside of its language's range.

Reply:
perhaps we should define a reasonable lower bound as a MUST? e.g. I don't think it makes sense to have conforming implementations that only support 256 element arrays

Reply:
Perhaps a signed 32 bit range could be mandatory (MUST) and a signed 64 bit range recommended (SHOULD).

Reply:
SGTM, w.r.t array length.

What about slice start and end positions? The RI offers an API that allows for full 64-bit absolute range *and* a full 64-bit range for "end relative" positions. If "end relative" positions are "naively" implemented with negative numbers (as suggested by the standard's own syntactic layer), that would limit implementations to only 63 bits to specify an absolute position (the same problem arises for those implementations that choose a 32-bit index space, and use 31-bit signed positions).

Reply:
On balance, I'd prefer to keep the spec simple rather than complicate it for "non-naive" implementations. Signed 64 bit values of slice start and end positions should be fine for all real-world JSON arrays. Implementations that choose to support only signed 32 bit values could be at a disadvantage for very large JSON arrays, but in that case, it would probably be better to use an implementation that supports signed 64 bit values.

Reply:
PTAL